### ISSUE TITLE
[1.8.0.2] Inability to import variables from a package

### ISSUE THREAD

**Issue by theodorDiaconu** [meteor/meteor#10409]
## Description
If I import from 'mobx' anything it fails like this:

```
modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:478 Uncaught RangeError: Maximum call stack size exceeded
    at modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:478
    at Array.every (<anonymous>)
    at fileAppendId (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:478)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:503)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
    at fileResolve (modules-runtime.js?hash=d3c3e5d67c95f97a60888bda7373292efad3be5e:562)
```

## Reproduction
```
git clone https://github.com/theodorDiaconu/meteor-fileResolve-infinite
cd meteor-fileResolve-infinite
meteor npm install
meteor run --port 4010
```

http://localhost:4010 , check console

**Comment by theodorDiaconu** [comment]
Problem I found happens here:

```
function fileResolve(file, id, parentModule, seenDirFiles) {
    var parentModule = parentModule || file.module;
    var extensions = fileGetExtensions(file);

    file =
      // Absolute module identifiers (i.e. those that begin with a `/`
      // character) are interpreted relative to the root directory, which
      // is a slight deviation from Node, which has access to the entire
      // file system.
      id.charAt(0) === "/" ? fileAppendId(root, id, extensions) :
      // Relative module identifiers are interpreted relative to the
      // current file, naturally.
      id.charAt(0) === "." ? fileAppendId(file, id, extensions) :
      // Top-level module identifiers are interpreted as referring to
      // packages in `node_modules` directories.
      nodeModulesLookup(file, id, extensions);

    // If the identifier resolves to a directory, we use the same logic as
    // Node to find an `index.js` or `package.json` file to evaluate.
    while (fileIsDirectory(file)) {
      seenDirFiles = seenDirFiles || [];

      // If the "main" field of a `package.json` file resolves to a
      // directory we've already considered, then we should not attempt to
      // read the same `package.json` file again. Using an array as a set
      // is acceptable here because the number of directories to consider
      // is rarely greater than 1 or 2. Also, using indexOf allows us to
      // store File objects instead of strings.
      if (seenDirFiles.indexOf(file) < 0) {
        seenDirFiles.push(file);

        var pkgJsonFile = fileAppendIdPart(file, "package.json"), main;
        var pkg = pkgJsonFile && fileEvaluate(pkgJsonFile, parentModule);
        if (pkg &&
            mainFields.some(function (name) {
              return isString(main = pkg[name]);
            })) {
          // The "main" field of package.json does not have to begin with
          // ./ to be considered relative, so first we try simply
          // appending it to the directory path before falling back to a
          // full fileResolve, which might return a package from a
          // node_modules directory.
          var mainFile = fileAppendId(file, main, extensions) ||
            fileResolve(file, main, parentModule, seenDirFiles);

          if (mainFile) {
            file = mainFile;
            recordChild(parentModule, pkgJsonFile);
            // The fileAppendId call above may have returned a directory,
            // so continue the loop to make sure we resolve it to a
            // non-directory file.
            continue;
          }
        }
      }

      // If we didn't find a `package.json` file, or it didn't have a
      // resolvable `.main` property, the only possibility left to
      // consider is that this directory contains an `index.js` module.
      // This assignment almost always terminates the while loop, because
      // there's very little chance `fileIsDirectory(file)` will be true
      // for `fileAppendIdPart(file, "index", extensions)`. However, in
      // principle it is remotely possible that a file called `index.js`
      // could be a directory instead of a file.
      file = fileAppendIdPart(file, "index", extensions);
    }

    if (file && isString(file.contents)) {
      file = fileResolve(file, file.contents, parentModule, seenDirFiles);
    }

    recordChild(parentModule, file);

    return file;
  };
```

The file contents is: `/node_modules/mobx/lib/mobx.js` so everytime it goes into `file = fileResolve(file, file.contents, parentModule, seenDirFiles);`

No idea how/why this happens.

**Comment by benjamn** [comment]
This seems to be because `mobx` has this useless self-referential `browser` [field](https://github.com/mobxjs/mobx/blob/c74c9f823c31ca20e977acd10305c625ee36a458/package.json#L9) in its `package.json`:
```json
  "browser": {
    "./lib/mobx.js": "./lib/mobx.js",
    "./lib/mobx.module.js": "./lib/mobx.module.js"
  },
```
Meteor supports the `browser` field by aliasing the replaced module to the replacement module, in which case `file.contents` is a string module identifier referring to the replacement. Hence the infinite loop.

**Comment by benjamn** [comment]
Oh geez, this no-op `browser` field was added just because it seemed to trick Webpack into doing the right thing, without any apparent understanding of why it helped, or what the unintended consequences might be: https://github.com/mobxjs/mobx/pull/1834

**Comment by benjamn** [comment]
@theodorDiaconu This should be fixed if you run `meteor update 1.8.1-beta.12`.

**Comment by theodorDiaconu** [comment]
Thank you @benjamn

**Comment by clhaas** [comment]
@benjamn latest version of MobX runs smoothly on Meteor 1.8.1-beta.12

**Comment by elie222** [comment]
> @theodorDiaconu This should be fixed if you run `meteor update 1.8.1-beta.12`.

For others, there's a small typo. Should be: `meteor update --release 1.8.1-beta.12`
